### PR TITLE
PR from develop to master

### PR DESCRIPTION
**Pull request by YuKopilets**
Rest api basics task (module2):
https://github.com/mjc-school/MJC-School/blob/master/modules/module%20%232.%20REST%20API%20Basics/rest_api_basics_task.md

### CODE REVIEW

**Review by maximelyashevich — CHANGES_REQUESTED** [review]
Please review my suggestions.

Regards,
Maks

File: GiftCertificatesSystem/service/src/test/java/com/epam/esm/service/impl/GiftCertificateServiceImplTest.java [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+
+@ExtendWith(SpringExtension.class)
+@ContextConfiguration(classes = TestConfig.class)
+@DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
+@ActiveProfiles("test")
+class GiftCertificateServiceImplTest {
+    @Autowired
+    private GiftCertificateDao giftCertificateDao;
+    @Autowired
+    private GiftCertificateService giftCertificateService;
```
Comment (maximelyashevich) [inline GiftCertificatesSystem/service/src/test/java/com/epam/esm/service/impl/GiftCertificateServiceImplTest.java:36]: I have a proposal: what if apply test configuration like that?

```Java
    private GiftCertificateService giftCertificateService;

    @Mock
    private GiftCertificateDao giftCertificateDao;

    @BeforeEach
    void setUp() {
        MockitoAnnotations.initMocks(this);

        giftCertificateService = spy(
                new GiftCertificateServiceImpl(giftCertificateDao)
        );
    }

    @AfterEach
    void tearDown() {
        giftCertificateDao = null;
        giftCertificateService = null;
    }
```

File: GiftCertificatesSystem/service/src/test/java/com/epam/esm/service/impl/GiftCertificateServiceImplTest.java [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+@ExtendWith(SpringExtension.class)
+@ContextConfiguration(classes = TestConfig.class)
+@DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
+@ActiveProfiles("test")
```
Comment (maximelyashevich) [inline GiftCertificatesSystem/service/src/test/java/com/epam/esm/service/impl/GiftCertificateServiceImplTest.java:31]: Do we really need id?

File: GiftCertificatesSystem/service/src/test/java/com/epam/esm/service/impl/GiftCertificateServiceImplTest.java [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+class GiftCertificateServiceImplTest {
+    @Autowired
+    private GiftCertificateDao giftCertificateDao;
+    @Autowired
+    private GiftCertificateService giftCertificateService;
+
+    @Test
+    void addGiftCertificateTest() {
```
Comment (maximelyashevich) [inline GiftCertificatesSystem/service/src/test/java/com/epam/esm/service/impl/GiftCertificateServiceImplTest.java:39]: Can you parameterize this method?

Please see `GiftCertificateDaoImplTest` for more details.

File: GiftCertificatesSystem/service/src/test/java/com/epam/esm/service/impl/GiftCertificateServiceImplTest.java [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+        List<GiftCertificate> actualGiftCertificates = giftCertificateService.getAllGiftCertificates();
+        assertNotEquals(exceptedGiftCertificates, actualGiftCertificates);
+    }
+
+    @Test
+    void updateGiftCertificateTest() throws ServiceException {
+        Optional<GiftCertificate> giftCertificate = Optional.ofNullable(createGiftCertificate());
+        Mockito.when(giftCertificateDao.findById(1L)).thenReturn(giftCertificate);
+        giftCertificateService.updateGiftCertificate(giftCertificate.get());
+        Mockito.verify(giftCertificateDao).update(giftCertificate.get());
```
Comment (maximelyashevich) [inline GiftCertificatesSystem/service/src/test/java/com/epam/esm/service/impl/GiftCertificateServiceImplTest.java:81]: Do not ignore NPE warning for `giftCertificate.get()`.

Please see `GiftCertificateDaoImplTest ` for more details.

File: GiftCertificatesSystem/service/src/test/java/com/epam/esm/service/impl/TagServiceImplTest.java [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+import java.util.List;
+import java.util.Optional;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+@ExtendWith(SpringExtension.class)
+@ContextConfiguration(classes = TestConfig.class)
+@DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
+@ActiveProfiles("test")
+class TagServiceImplTest {
```
Comment (maximelyashevich) [inline GiftCertificatesSystem/service/src/test/java/com/epam/esm/service/impl/TagServiceImplTest.java:27]: Please check recommendations from `GiftCertificateServiceImplTest` and try to apply them for `TagServiceImplTest`.

File: GiftCertificatesSystem/web/src/main/java/com/epam/esm/controller/GiftCertificateController.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+        return giftCertificateService.sortGiftCertificatesByDateAsc();
+    }
+
+    @GetMapping(value = "/sort/date/desc")
+    public List<GiftCertificate> getGiftCertificatesByDateDesc() {
+        return giftCertificateService.sortGiftCertificatesByDateDesc();
+    }
+
+    @PutMapping
+    public GiftCertificate updateGiftCertificate(@RequestBody GiftCertificate giftCertificate) {
```
Comment (maximelyashevich) [inline GiftCertificatesSystem/web/src/main/java/com/epam/esm/controller/GiftCertificateController.java:103]: Does it make sense to use `/certificates/{id}` endpoint?

**Review by maximelyashevich — CHANGES_REQUESTED** [review]
Please review my recommendations. Apart from that, check H2 database schema and verify that all persistence tests are working.

Do not forget that service projects should be covered not less than 80% tests. 

Regards,
Maks

File: GiftCertificatesSystem/persistence/src/test/java/com/epam/esm/dao/impl/GiftCertificateDaoImplTest.java [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+        LocalDateTime localDateTime = LocalDateTime.now();
+        return GiftCertificate.builder()
+                .name("test name")
+                .description("test description")
+                .price(BigDecimal.valueOf(20.30).setScale(2, RoundingMode.HALF_UP))
+                .createDate(localDateTime)
+                .lastUpdateDate(localDateTime)
+                .duration(25)
+                .build();
+    }
```
Comment (maximelyashevich) [inline GiftCertificatesSystem/persistence/src/test/java/com/epam/esm/dao/impl/GiftCertificateDaoImplTest.java:101]: Please consider `DataProviderUtils` class which can generate the required data.

File: GiftCertificatesSystem/persistence/src/test/java/com/epam/esm/dao/impl/TagDaoImplTest.java [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+                .id(5L)
+                .name("fifthTag")
+                .build();
+        tags.add(firstTag);
+        tags.add(secondTag);
+        tags.add(thirdTag);
+        tags.add(fourthTag);
+        tags.add(fifthTag);
+        return tags;
+    }
```
Comment (maximelyashevich) [inline GiftCertificatesSystem/persistence/src/test/java/com/epam/esm/dao/impl/TagDaoImplTest.java:134]: Please consider `DataProviderUtils` class which can generate the required data.

File: GiftCertificatesSystem/service/src/main/java/com/epam/esm/service/GiftCertificateService.java [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    List<GiftCertificate> getAllGiftCertificates();
+
+    List<GiftCertificate> getGiftCertificatesByQueryParams(GiftCertificateQuery giftCertificateQuery);
+
+    GiftCertificate updateGiftCertificate(GiftCertificate giftCertificate)
+            throws GiftCertificateNotFoundServiceException;
+
+    void removeGiftCertificate(Long id) throws InvalidRequestedIdServiceException;
+
+    void addGiftCertificateTags(GiftCertificate giftCertificate);
```
Comment (maximelyashevich) [inline GiftCertificatesSystem/service/src/main/java/com/epam/esm/service/GiftCertificateService.java:25]: Please check method name and make it more clear.

File: GiftCertificatesSystem/service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+    public GiftCertificate getGiftCertificateById(Long id)
+            throws GiftCertificateNotFoundServiceException, InvalidRequestedIdServiceException {
+        if (id > 0) {
+            Optional<GiftCertificate> giftCertificateById = giftCertificateDao.findById(id);
+            return giftCertificateById.orElseThrow(() ->
+                    new GiftCertificateNotFoundServiceException("GiftCertificate with id=" + id + " not found")
+            );
+        } else {
+            throw new InvalidRequestedIdServiceException(id + " does not fit the allowed gap. Expected gap: 0 > id");
+        }
```
Comment (maximelyashevich) [inline GiftCertificatesSystem/service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java:45]:
```suggestion
        if (id <= 0) {
            throw new InvalidRequestedIdServiceException(id + " does not fit the allowed gap. Expected gap: 0 > id");
        }
         
        Optional<GiftCertificate> giftCertificateById = giftCertificateDao.findById(id);
        return giftCertificateById.orElseThrow(() ->
                    new GiftCertificateNotFoundServiceException("GiftCertificate with id=" + id + " not found")
            );
```

Please consider validation method to check id.

File: GiftCertificatesSystem/service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+            }
+            if (giftCertificate.getDescription() != null) {
+                giftCertificateById.setDescription(giftCertificate.getDescription());
+            }
+            if (giftCertificate.getPrice() != null) {
+                giftCertificateById.setPrice(giftCertificate.getPrice());
+            }
+            if (giftCertificate.getDuration() != null) {
+                giftCertificateById.setDuration(giftCertificate.getDuration());
+            }
```
Comment (maximelyashevich) [inline GiftCertificatesSystem/service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java:78]: Try to optimize the number of conditional statements.

File: GiftCertificatesSystem/service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+    @Override
+    @Transactional
+    public void removeGiftCertificate(Long id) throws InvalidRequestedIdServiceException {
+        if (id > 0) {
+            giftCertificateDao.delete(id);
+            removeGiftCertificateTags(id);
+        } else {
+            throw new InvalidRequestedIdServiceException(id + " does not fit the allowed gap. Expected gap: 0 > id");
```
Comment (maximelyashevich) [inline GiftCertificatesSystem/service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java:94]: Please consider validation method to check id - the idea is not to have code duplication.

File: GiftCertificatesSystem/service/src/main/java/com/epam/esm/service/impl/TagServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+import com.epam.esm.service.exception.InvalidRequestedIdServiceException;
+import com.epam.esm.service.exception.TagNotFoundServiceException;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.List;
+import java.util.Optional;
+
+@Service
+public class TagServiceImpl implements TagService {
```
Comment (maximelyashevich) [inline GiftCertificatesSystem/service/src/main/java/com/epam/esm/service/impl/TagServiceImpl.java:15]: Please check my recommendations for `GiftSertificateServiceImpl` class.

File: GiftCertificatesSystem/web/src/main/java/com/epam/esm/controller/GiftCertificateController.java [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+        return giftCertificateService.addGiftCertificate(giftCertificate);
+    }
+
+    @GetMapping(value = "/{id}")
+    public GiftCertificate getGiftCertificateById(@PathVariable long id) {
+        try {
+            return giftCertificateService.getGiftCertificateById(id);
+        } catch (GiftCertificateNotFoundServiceException e) {
+            log.error("Failed to get gift certificate by id (gift certificate not found)", e);
+            throw new GiftCertificateNotFoundResponseStatusException(e);
```
Comment (maximelyashevich) [inline GiftCertificatesSystem/web/src/main/java/com/epam/esm/controller/GiftCertificateController.java:45]: Please check if it is a good practice to use log + throw.

File: GiftCertificatesSystem/web/src/main/java/com/epam/esm/controller/TagController.java [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.DeleteMapping;
+
+import java.util.List;
+
+@Slf4j
+@RestController
+@RequestMapping("/tags")
+public class TagController {
```
Comment (maximelyashevich) [inline GiftCertificatesSystem/web/src/main/java/com/epam/esm/controller/TagController.java:23]: Please check my recommendations for `GiftCertificateController` class.

File: GiftCertificatesSystem/web/src/main/java/com/epam/esm/exception/GiftCertificateNotFoundResponseStatusException.java [hunk 16]
```diff
@@ -0,0 +1,22 @@
+package com.epam.esm.exception;
+
+import org.springframework.http.HttpStatus;
+import org.springframework.web.server.ResponseStatusException;
+
+public class GiftCertificateNotFoundResponseStatusException extends ResponseStatusException {
```
Comment (maximelyashevich) [inline GiftCertificatesSystem/web/src/main/java/com/epam/esm/exception/GiftCertificateNotFoundResponseStatusException.java:6]: What's about `ExceptionHandler` class?

Please weigh up all pros and cons to use the current approach.

**Review by maximelyashevich — CHANGES_REQUESTED** [review]

File: GiftCertificatesSystem/service/src/main/java/com/epam/esm/service/GiftCertificateService.java (hunk 9 above)
Comment (YuKopilets) [reply GiftCertificatesSystem/service/src/main/java/com/epam/esm/service/GiftCertificateService.java:25]: Can you place me any example how better name them?

File: GiftCertificatesSystem/persistence/src/test/java/com/epam/esm/dao/impl/GiftCertificateDaoImplTest.java [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+
+    @Test
+    void findAllGiftCertificatesTest() {
+        List<GiftCertificate> expectedGiftCertificates = createExpectedGiftCertificates();
+        List<GiftCertificate> actualGiftCertificates = giftCertificateDao.findAll();
+        assertEquals(expectedGiftCertificates, actualGiftCertificates);
+    }
+
+    @Test
+    void findAllGiftCertificatesNegativeTest() {
```
Comment (MaksimShakavin) [inline GiftCertificatesSystem/persistence/src/test/java/com/epam/esm/dao/impl/GiftCertificateDaoImplTest.java:64]: I like using "should" in tests naming. It makes test more like documentation of the class, and more tdd like. Also you may not rename the test method when you rename the origina method. I.e shouldReturnEmptyWhenNoCertificate, shouldDeleteCertificate, shouldReturnAllCertificated

File: GiftCertificatesSystem/service/src/main/java/com/epam/esm/service/exception/ServiceException.java [hunk 18]
```diff
@@ -0,0 +1,19 @@
+package com.epam.esm.service.exception;
+
+public class ServiceException extends Exception {
```
Comment (MaksimShakavin) [inline GiftCertificatesSystem/service/src/main/java/com/epam/esm/service/exception/ServiceException.java:3]: Why not runtimeException? holy war topic

File: GiftCertificatesSystem/service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+
+import java.time.LocalDateTime;
+import java.util.List;
+import java.util.Optional;
+
+@Service
+public class GiftCertificateServiceImpl implements GiftCertificateService {
+    private final GiftCertificateDao giftCertificateDao;
+
+    public GiftCertificateServiceImpl(GiftCertificateDao giftCertificateDao) {
```
Comment (MaksimShakavin) [inline GiftCertificatesSystem/service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java:20]: can we use lombok here as well?

File: GiftCertificatesSystem/web/src/main/java/com/epam/esm/controller/GiftCertificateController.java [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+        } catch (InvalidRequestedIdServiceException e) {
```
Comment (MaksimShakavin) [inline GiftCertificatesSystem/web/src/main/java/com/epam/esm/controller/GiftCertificateController.java:46]: I would better catch one ServiceException and provide an appropriate messages in it's descendants

**Review by MaksimShakavin — CHANGES_REQUESTED** [review]

File: GiftCertificatesSystem/service/src/main/java/com/epam/esm/service/GiftCertificateService.java (hunk 9 above)
Comment (maximelyashevich) [reply GiftCertificatesSystem/service/src/main/java/com/epam/esm/service/GiftCertificateService.java:25]: For example, `addTagsToGiftCerificate()` or `assignTagsToGiftCerificate()`.

File: GiftCertificatesSystem/persistence/src/main/resources/datasource.properties [hunk 21]
```diff
@@ -0,0 +1,4 @@
+dataSource.driverClassName=com.mysql.cj.jdbc.Driver
+dataSource.url=jdbc:mysql://localhost:3306/gift_certificates_system?useUnicode=true&serverTimezone=UTC
+dataSource.username=root
+dataSource.password=root
```
[the remark this answers is not in the document]
Comment (YuKopilets) [reply GiftCertificatesSystem/persistence/src/main/resources/datasource.properties:4]: How can i do it better?

File: GiftCertificatesSystem/persistence/src/main/java/com/epam/esm/dao/impl/TagDaoImpl.java [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+            tag.setId(key.longValue());
+        }
+        return tag;
+    }
+
+    @Override
+    public Optional<Tag> findById(Long id) {
+        try {
+            return Optional.ofNullable(jdbcTemplate.queryForObject(SELECT_TAG, tagMapper, id));
+        } catch (EmptyResultDataAccessException e) {
```
[the remark this answers is not in the document]
Comment (YuKopilets) [reply GiftCertificatesSystem/persistence/src/main/java/com/epam/esm/dao/impl/TagDaoImpl.java:52]: RowMapper doesn't return null if db hasn't got a record with actual id. It generate EmptyResultDataAccessException

File: GiftCertificatesSystem/service/src/test/java/com/epam/esm/service/impl/GiftCertificateServiceImplTest.java [hunk 23]
```diff
@@ -0,0 +1,172 @@
...
+        List<GiftCertificate> actualGiftCertificates = giftCertificateService.getGiftCertificates(giftCertificateQuery);
+        assertNotEquals(exceptedGiftCertificates, actualGiftCertificates);
+    }
+
+    @ParameterizedTest
+    @MethodSource("prepareGiftCertificate")
+    void updateGiftCertificateTest(GiftCertificate giftCertificate) throws ServiceException {
+        Optional<GiftCertificate> giftCertificateOptional = Optional.ofNullable(giftCertificate);
+        Mockito.when(giftCertificateDao.findById(1L)).thenReturn(giftCertificateOptional);
+        if (giftCertificateOptional.isPresent()) {
```
Comment (maximelyashevich) [inline GiftCertificatesSystem/service/src/test/java/com/epam/esm/service/impl/GiftCertificateServiceImplTest.java:90]: You can also use `Optional.ifPresent()` method.

File: GiftCertificatesSystem/web/src/main/java/com/epam/esm/error/ControllerExceptionHandler.java [hunk 24]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    @ExceptionHandler({DeleteByRequestedIdServiceException.class,
+            GiftCertificateNotFoundServiceException.class,
+            TagNotFoundServiceException.class})
+    public ResponseEntity<Object> handleServiceException(ServiceException ex, WebRequest request) {
+        return handleException(ex, HttpStatus.BAD_REQUEST);
+    }
+
+    @ExceptionHandler(RuntimeException.class)
```
Comment (maximelyashevich) [inline GiftCertificatesSystem/web/src/main/java/com/epam/esm/error/ControllerExceptionHandler.java:33]: Can we consider here `Exception` class?

**Review by maximelyashevich — CHANGES_REQUESTED** [review]
A few minor changes are required.

File: GiftCertificatesSystem/web/src/main/java/com/epam/esm/error/ControllerExceptionHandler.java (hunk 24 above)
Comment (YuKopilets) [reply GiftCertificatesSystem/web/src/main/java/com/epam/esm/error/ControllerExceptionHandler.java:33]: We discussed that only RuntimeExceptions are used in the spring

File: GiftCertificatesSystem/service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 25]
```diff
@@ -0,0 +1,141 @@
...
+                || StringUtils.isNotEmpty(giftCertificateQuery.getSort())
+        ) {
+            String sort = giftCertificateQuery.getSort();
+            if (StringUtils.isNotEmpty(sort) && !("name".equals(sort) || "date".equals(sort))) {
+                giftCertificateQuery.setSort(null);
+            }
+            return true;
+        } else {
+            return false;
+        }
```
[the remark this answers is not in the document]
Comment (YuKopilets) [reply GiftCertificatesSystem/service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java:104]: You said me that we can add else block for write more readable code

File: GiftCertificatesSystem/service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 26]
```diff
@@ -0,0 +1,141 @@
...
+        );
+    }
+
+    @Override
+    public List<GiftCertificate> getGiftCertificates(GiftCertificateQuery giftCertificateQuery) {
+        if (reviewGiftCertificateQueryParams(giftCertificateQuery)) {
+            return giftCertificateDao.findAllByQueryParams(giftCertificateQuery);
+        } else {
+            return giftCertificateDao.findAll();
+        }
```
[the remark this answers is not in the document]
Comment (YuKopilets) [reply GiftCertificatesSystem/service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java:51]: You said me that we can add else block for write more readable code

**Review by maximelyashevich — APPROVED** [review]

**Pull request merged**